### ISSUE TITLE
Bug: Grpc providers don't respect filepath scope from chained conditions

### ISSUE THREAD

**Issue by tsanders-rh** [konveyor/analyzer-lsp#1002]
# Bug: Grpc providers don't respect filepath scope from chained conditions

## Summary

When using rule chaining with `from:` to pass filepaths from a previous condition to a grpc provider (nodejs.referenced, java.referenced, etc.), the grpc provider ignores the filepath scope and searches the entire codebase instead. This makes it impossible to use grpc providers in chained conditions to filter analysis to specific files.

## Related Issues

This is a separate issue from #1000 (template variables not populated). Even if #1000 is fixed, chaining with grpc providers still won't work correctly due to this filepath scoping bug.

## Environment

- **analyzer-lsp**: Latest version (tested 2024-11)
- **Affects**: All grpc providers (nodejs.referenced, java.referenced, etc.)
- **Works correctly**: builtin providers (builtin.filecontent, builtin.file, etc.)

### Test Rule (Chaining Within Single Rule)

```yaml
- ruleID: toolbar-group-scoped-check
  when:
    and:
    # Step 1: Find files that import ToolbarGroup
    - as: importFiles
      nodejs.referenced:
        pattern: ToolbarGroup
        location: ImportDeclaration
    # Step 2: Check ONLY those files for JSX usage (should be scoped)
    - from: importFiles.Filepaths  # ← Should limit search to files from step 1
      nodejs.referenced:
        pattern: ToolbarGroup
        location: JSXOpeningElement
  message: "Found ToolbarGroup usage in: {{file}}"
```

### Expected Behavior

The second condition (nodejs.referenced) should **only** analyze the files identified in the first condition (e.g., 5 files that import ToolbarGroup).

### Actual Behavior

The second condition **receives no filepaths** and produces zero matches, making chaining completely broken.

### Evidence

When we tested this, we saw:
- First condition: nodejs.referenced finds imports (but doesn't populate filepaths - see issue #1000)
- Second condition: receives empty filepath list from `from:`, resulting in 0 matches
- When nodejs.referenced runs **standalone** (without chaining), it finds 17+ matches across all files
- But in a chain, both bugs (#1000 and #1002) combine to make it completely non-functional

## Root Cause

**Location**: `provider/provider.go:357-378`

The grpc provider code receives the filepath scope from `ProviderContext.Template[TemplateContextPathScopeKey]` but doesn't use it to filter the analysis:

```go
func (p *grpcProviderClient) ProviderEvaluateRuleCond(
    ctx context.Context,
    cap string,
    conditionInfo ProviderCondition,
) (ProviderEvaluateResponse, error) {
    // ... code ...

    condCtx := &proto.ConditionContext{}
    if value, ok := p.Template[engine.TemplateContextPathScopeKey]; ok {
        b, err := json.Marshal(value)
        if err != nil {
            return response, err
        }
        condCtx.TemplateContext = map[string]string{
            engine.TemplateContextPathScopeKey: string(b),
        }
    }
    // ← Code passes filepaths to grpc provider...

    conditionRPC := &proto.Condition{
        ConditionInfo: string(b),
        Context:       condCtx,
    }
    resp, err := p.client.Evaluate(ctx, conditionRPC)
    // ← But grpc provider doesn't actually use them to filter!
}
```

The filepath scope is serialized and passed to the grpc provider, but **the grpc provider implementations don't respect this constraint**. They receive the scope data but continue to search all files.

### builtin.filecontent (WORKS ✅)

The builtin provider correctly filters based on filepath scope:

```go
// In builtin provider
if len(filepaths) > 0 {
    // Only search these specific files
    for _, filepath := range filepaths {
        searchFile(filepath, pattern)
    }
} else {
    // Search all files
    searchAllFiles(pattern)
}
```

### nodejs.referenced (BROKEN ❌)

The nodejs grpc provider ignores the filepath scope and always searches everything:

```go
// In nodejs provider (analyzer-lsp-grpc)
func (p *nodejsProvider) Evaluate(ctx context.Context, cond *Condition) (*Response, error) {
    // Receives condCtx.TemplateContext with filepaths...
    // But doesn't use them!

    // Always searches entire workspace:
    return p.searchAllFiles(pattern)
}
```

## Impact

This bug affects:
- ✅ **All** chained conditions using grpc providers with `from:`
- ✅ **All** grpc providers (nodejs.referenced, java.referenced, dotnet.referenced, etc.)
- ✅ Performance (searches entire codebase when it should search a few files)
- ✅ Correctness (finds matches in files that don't meet the chain criteria)

### Use Cases That Don't Work

1. **Import verification chains** (within single rule):
   ```yaml
   - ruleID: import-verification-chain
     when:
       and:
       # Find files that import Component X
       - as: importFiles
         nodejs.referenced:
           pattern: ComponentX
           location: ImportDeclaration
       # Check ONLY those files for usage (BROKEN - searches all files)
       - from: importFiles.Filepaths
         nodejs.referenced:
           pattern: ComponentX
           location: JSXOpeningElement
   ```

2. **Scope-limited refactoring** (within single rule):
   ```yaml
   - ruleID: api-refactoring-check
     when:
       and:
       # Find files using deprecated API
       - as: apiFiles
         nodejs.referenced: { pattern: OldAPI }
       # Check ONLY those files for new API (BROKEN - searches all files)
       - from: apiFiles.Filepaths
         nodejs.referenced: { pattern: NewAPI }
   ```

**Note**: The Java provider has a post-filtering workaround for filepath scoping, so `java.referenced → java.referenced` chaining may work better than other grpc providers. However, the proper fix would avoid the unnecessary analysis of files that get filtered out.

### Option 1: Implement Filepath Filtering in Grpc Providers

Modify grpc provider implementations (in analyzer-lsp repository and provider repositories like konveyor/analyzer-lsp) to:

1. Extract filepaths from `condCtx.TemplateContext[TemplateContextPathScopeKey]`
2. When filepaths are present, limit analysis to only those files
3. When filepaths are empty, search all files (current behavior)

**Example implementation** (pseudocode for nodejs provider):

```go
func (p *nodejsProvider) Evaluate(ctx context.Context, cond *Condition) (*Response, error) {
    var filepaths []string

    // Extract filepath scope if present
    if scopeData, ok := cond.Context.TemplateContext[TemplateContextPathScopeKey]; ok {
        var scope ChainTemplate
        json.Unmarshal([]byte(scopeData), &scope)
        filepaths = scope.Filepaths
    }

    // Search files
    if len(filepaths) > 0 {
        // Only analyze the specified files
        return p.searchFiles(filepaths, pattern)
    } else {
        // Search all files (current behavior)
        return p.searchAllFiles(pattern)
    }
}
```

### Option 2: Filter Results in Engine After Provider Returns

Alternatively, modify `engine/conditions.go` to filter provider results based on filepath scope:

```go
response, err := c.Evaluate(ctx, log, condCtx)
if err != nil {
    return ConditionResponse{}, err
}

// If we have a filepath scope, filter incidents to only those files
if scopeValue, ok := condCtx.Template[TemplateContextPathScopeKey]; ok {
    allowedFiles := map[string]bool{}
    for _, fp := range scopeValue.Filepaths {
        allowedFiles[fp] = true
    }

    filteredIncidents := []IncidentContext{}
    for _, incident := range response.Incidents {
        if allowedFiles[incident.FileURI.Filename()] {
            filteredIncidents = append(filteredIncidents, incident)
        }
    }
    response.Incidents = filteredIncidents
}
```

**Option 1 is preferred** because it avoids unnecessary work (searching files that will be filtered out anyway).

## Workaround

The only current workaround is to use **builtin.filecontent** for both conditions (since builtin providers respect filepath scoping):

```yaml
- ruleID: toolbar-group-migration
  when:
    and:
    # Step 1: Find files with imports using regex
    - as: importFiles
      builtin.filecontent:
        pattern: "import.*ToolbarGroup.*@patternfly/react-core"
    # Step 2: Search ONLY those files for usage
    - from: importFiles  # ✅ builtin respects filepath scoping
      builtin.filecontent:
        pattern: "<ToolbarGroup"
  message: "Found ToolbarGroup usage in {{file}} - migrate to ToolbarItem"
```

This workaround:
- ✅ Works correctly with filepath scoping
- ✅ More efficient (only searches relevant files in step 2)
- ❌ Can't use nodejs.referenced for semantic analysis (has to use regex patterns)
- ❌ Less precise than AST-based detection
- ❌ May miss complex import patterns or have false positives

**Note**: Using `nodejs.referenced → builtin.filecontent` does NOT work because nodejs.referenced doesn't populate the filepaths (see issue #1000), so the second condition has no files to scope to.

## Test Case

See test file demonstrating the bug (chaining within a single rule):

<details>
<summary>test-filepath-scope.yaml</summary>

```yaml
- name: test-filepath-scope
  description: Test case demonstrating grpc provider filepath scope bug
  rules:
  # Chaining within single rule with and block
  - ruleID: test-filepath-scoping
    when:
      and:
      # Step 1: Find files with imports (should find ~5 files)
      - as: scopedFiles
        nodejs.referenced:
          pattern: ToolbarGroup
          location: ImportDeclaration
      # Step 2: Search ONLY those files for usage (but grpc provider ignores scope)
      - from: scopedFiles.Filepaths  # ← Should limit to ~5 files
        nodejs.referenced:
          pattern: ToolbarGroup
          location: JSXOpeningElement
    message: "Should only find matches in files from step 1, but finds matches everywhere"
```
</details>

### Expected Output
```
~5-10 matches (only in files that import ToolbarGroup)
```

### Actual Output
```
0 matches (chaining completely broken - no filepaths passed to second condition)
```

**Note**: When nodejs.referenced runs standalone (without chaining), it finds 17+ matches across all files. But in a chain, the second condition gets no filepaths due to bug #1000, resulting in zero matches.

## Additional Context

This issue was discovered while developing PatternFly v5 to v6 migration rules. The inability to chain grpc providers forces verbose workarounds and prevents leveraging the AST-based analysis capabilities of providers like nodejs.referenced.

### Related Code Locations

- `provider/provider.go:357-378` - Where filepaths are passed to grpc providers
- `engine/conditions.go:170-185` - Where filepath scope is set in ChainTemplate
- Provider implementations in analyzer-lsp repositories (nodejs, java, dotnet, etc.)

### Benefits of Fixing

1. Enable true semantic chaining (AST → import → usage)
2. Improve performance (search fewer files)
3. Improve accuracy (only match in relevant files)
4. Simplify complex migration rules
5. Match user expectations for `from:` behavior

## Related Issues

- #1000 - Template variables from chained conditions not populated (separate bug)

Together, fixing both #1000 and this issue would enable full rule chaining support for grpc providers.

**Comment by shawn-hurley** [comment]
The java provider also does a post filter on [filepaths](https://github.com/konveyor/analyzer-lsp/blob/main/external-providers/java-external-provider/pkg/java_external_provider/service_client.go#L186-L199) so while the analysis of that file will take the exact time, the results should be impacted. 

I think that we have too many different ways to include/exclude results, I think we need to get a handle on this so very much in favor of fixing this.

**Comment by tsanders-rh** [comment]
Thank you for the clarification, Shawn! You're absolutely right - I incorrectly showed chaining across separate rules in the examples. Chaining only works within `and:`/`or:` blocks in a single rule.

I also appreciate the insight about the Java provider already doing post-filtering as a workaround. That's helpful context.

I'll update this issue to correct the examples to show proper chaining within a single rule's condition blocks.
